relay-server/src/services/cogs.rs

```diff
@@ -0,0 +1,183 @@
+use std::sync::atomic::{AtomicBool, Ordering};
+
+use relay_cogs::{CogsMeasurement, CogsRecorder, ResourceId};
+use relay_config::Config;
+use relay_system::{Addr, Controller, FromMessage, Interface, Service};
+use sentry_usage_accountant::{Producer, UsageAccountant, UsageUnit};
+
+#[cfg(feature = "processing")]
+use crate::services::store::{Store, StoreCogs};
+
+pub struct CogsReport(CogsMeasurement);
+
+impl Interface for CogsReport {}
+
+impl FromMessage<CogsMeasurement> for CogsReport {
+    type Response = relay_system::NoResponse;
+
+    fn from_message(message: CogsMeasurement, _: ()) -> Self {
+        Self(message)
+    }
+}
+
+/// Relay [accountant producer](Producer) which produces to a dynamic
+/// upstram.
+enum RelayProducer {
+    /// Produces to Kafka via the store service.
+    #[cfg(feature = "processing")]
+    Store(Addr<Store>),
+    /// Discards all measurements.
+    Noop,
+}
+
+impl Producer for RelayProducer {
+    type Error = std::convert::Infallible;
+
+    fn send(&mut self, _payload: Vec<u8>) -> Result<(), Self::Error> {
+        match self {
+            #[cfg(feature = "processing")]
+            Self::Store(addr) => addr.send(StoreCogs(_payload)),
+            Self::Noop => {}
+        }
+
+        Ok(())
+    }
+}
+
+/// Service implementing the [`CogsReport`] interface.
+pub struct CogsService {
+    relay_resource_id: String,
+    usage_accountant: UsageAccountant<RelayProducer>,
+}
+
+impl CogsService {
+    pub fn new(config: &Config, #[cfg(feature = "processing")] store: Option<Addr<Store>>) -> Self {
+        #[cfg(feature = "processing")]
+        let producer = store
+            .map(RelayProducer::Store)
+            .filter(|_| config.cogs_enabled())
+            .unwrap_or(RelayProducer::Noop);
+
+        #[cfg(not(feature = "processing"))]
+        let producer = RelayProducer::Noop;
```

Comment:
Are there plans to extend the support for PoP relays too?

Reply:
Yes! One of the reasons the resource id is mapped from the config and just a constant.